File: spring-webflux/src/main/java/software/darkmatter/school/blog/domain/comment/business/CommentServiceImpl.java
package software.darkmatter.school.blog.domain.comment.business;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import software.darkmatter.school.blog.api.dto.CommentCreateDto;
import software.darkmatter.school.blog.api.dto.CommentUpdateDto;
import software.darkmatter.school.blog.domain.comment.data.Comment;
import software.darkmatter.school.blog.domain.comment.data.CommentRepository;
import software.darkmatter.school.blog.domain.comment.error.CommentNotFoundException;
import software.darkmatter.school.blog.domain.post.business.PostService;
import software.darkmatter.school.blog.domain.user.business.UserService;

import java.time.OffsetDateTime;

import static software.darkmatter.school.blog.config.security.SimpleAuthentication.simpleAuthFromContext;

@Service
@RequiredArgsConstructor
public class CommentServiceImpl implements CommentService {

    private final CommentRepository repository;
    private final UserService userService;
    private final PostService postService;

    @Override
    public Flux<Comment> getListByPostId(Long postId, Pageable pageable) {
        return postService.getById(postId).thenMany(
            repository.findAllByPostIdAndDeletedAtIsNull(postId, pageable)
        );
    }

    @Override
    public Mono<Comment> getById(Long id) {
        return repository.findByIdAndDeletedAtIsNull(id)
                         .switchIfEmpty(Mono.error(() -> new CommentNotFoundException(id)))
                         .flatMap(comment ->
                                      userService.getById(comment.getCreatedByUserId())
                                                 .map(
                                                     user -> {
                                                         comment.setCreatedBy(user);
                                                         return comment;
                                                     }
                                                 )
                         );
    }

    @Override
    @Transactional(readOnly = true)
    public Mono<Comment> getByPostIdAndId(Long postId, Long id) {
        return repository.findByPostIdAndIdAndDeletedAtIsNull(postId, id)
                         .switchIfEmpty(Mono.error(() -> new CommentNotFoundException(id)))
                         .flatMap(comment ->
                                      userService.getById(comment.getCreatedByUserId())
                                                 .map(
                                                     user -> {
                                                         comment.setCreatedBy(user);
                                                         return comment;
                                                     }
                                                 )
                         );
    }

    @Override
    public Mono<Comment> create(Long postId, CommentCreateDto createDto) {
        return simpleAuthFromContext().flatMap(
            authentication -> postService.getById(postId).flatMap(
                post -> userService.getById(authentication.getUserId()).flatMap(
                    user -> {
                        var comment = new Comment();
                        comment.setPostId(postId);
                        comment.setText(createDto.text());
                        OffsetDateTime now = OffsetDateTime.now();
                        comment.setCreatedAt(now);
                        comment.setCreatedByUserId(user.getId());
                        comment.setUpdatedAt(now);
                        comment.setUpdatedByUserId(user.getId());
                        return repository.save(comment);
                    })
            )
        );
    }

    @Override
    public Mono<Comment> update(Long id, CommentUpdateDto updateDto) {
        return simpleAuthFromContext().flatMap(
            authentication ->
                Mono.zip(
                    userService.getById(authentication.getUserId()),
                    getById(id)
                ).flatMap(pair -> {
                    var user = pair.getT1();
                    var comment = pair.getT2();

                    comment.setText(updateDto.text());
                    comment.setUpdatedAt(OffsetDateTime.now());
                    comment.setUpdatedByUserId(user.getId());
                    return repository.save(comment);
                })
        );
    }

    @Override
    public Mono<Void> delete(Long id) {
        return simpleAuthFromContext().flatMap(
            authentication ->
                Mono.zip(
                    userService.getById(authentication.getUserId()),
                    getById(id)
                ).flatMap((pair) -> {
                    var user = pair.getT1();
                    var comment = pair.getT2();

                    comment.setDeletedAt(OffsetDateTime.now());
                    comment.setDeletedByUserId(user.getId());
                    return repository.save(comment);
                }).then()
        );
    }
}

File: kotlin-spring-webflux/src/main/kotlin/software/darkmatter/school/blog/domain/comment/business/CommentServiceImpl.kt
package software.darkmatter.school.blog.domain.comment.business

import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.reactor.awaitSingle
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import software.darkmatter.school.blog.api.dto.CommentCreateDto
import software.darkmatter.school.blog.api.dto.CommentUpdateDto
import software.darkmatter.school.blog.config.security.SimpleAuthentication.Companion.simpleAuthFromContext
import software.darkmatter.school.blog.domain.comment.data.Comment
import software.darkmatter.school.blog.domain.comment.data.CommentRepository
import software.darkmatter.school.blog.domain.comment.error.CommentNotFoundException
import software.darkmatter.school.blog.domain.post.business.PostService
import software.darkmatter.school.blog.domain.user.business.UserService
import java.time.OffsetDateTime

@Service
class CommentServiceImpl(
    private val repository: CommentRepository,
    private val userService: UserService,
    private val postService: PostService
) : CommentService {

    override suspend fun getListByPostId(postId: Long, pageable: Pageable): List<Comment> =
        repository.findAllByPostIdAndDeletedAtIsNull(postId, pageable).toList()

    override suspend fun getById(id: Long): Comment {
        val comment = repository.findByIdAndDeletedAtIsNull(id) ?: throw CommentNotFoundException(id)
        comment.createdBy = userService.getById(comment.createdByUserId)

        return comment
    }

    override suspend fun getByPostIdAndId(postId: Long, id: Long): Comment {
        val comment = repository.findByPostIdAndIdAndDeletedAtIsNull(postId, id) ?: throw CommentNotFoundException(id)
        comment.createdBy = userService.getById(comment.createdByUserId)

        return comment
    }

    override suspend fun create(postId: Long, createDto: CommentCreateDto): Comment {
        postService.getById(postId)
        val authentication = simpleAuthFromContext()
        val user = userService.getById(authentication.userId)
        val createdAt = OffsetDateTime.now()

        val comment = Comment(
            postId = postId,
            text = createDto.text!!,
            createdAt = createdAt,
            createdByUserId = user.id!!,
            updatedAt = createdAt,
            updatedByUserId = user.id,
            deletedAt = null,
            deletedByUserId = null
        )

        return repository.save(comment).awaitSingle()
    }

    override suspend fun update(id: Long, updateDto: CommentUpdateDto): Comment {
        val authentication = simpleAuthFromContext()
        val user = userService.getById(authentication.userId)
        val comment = getById(id)

        comment.text = updateDto.text!!
        comment.updatedAt = OffsetDateTime.now()
        comment.updatedByUserId = user.id!!

        return repository.save(comment).awaitSingle()
    }

    override suspend fun delete(id: Long) {
        val authentication = simpleAuthFromContext()
        val user = userService.getById(authentication.userId)
        val comment = getById(id)

        comment.deletedAt = OffsetDateTime.now()
        comment.deletedByUserId = user.id!!

        repository.save(comment).awaitSingle()
    }
}
